nearcore/src/config.rs

```diff
@@ -1108,20 +1112,223 @@ pub enum FileDownloadError {
+    /// decompressing it if the stream is XZ-compressed.  Note that once all the
+    /// data has been written [`finish`] function must be called to flush
+    /// internal buffers.
+    async fn write_all(&mut self, chunk: &[u8]) -> Result<(), FileDownloadError> {
+        if let Some(len) = self.probe(chunk) {
+            if len != 0 {
+                self.write_all_impl(&XZ_HEADER_MAGIC[..len]).await?;
+            }
+            self.write_all_impl(&chunk).await?;
+        }
+        Ok(())
+    }
+
+    /// Flushes all internal buffers and closes the output file.
+    async fn finish(mut self) -> Result<(), FileDownloadError> {
+        match self.state {
+            AutoXzState::Probing(pos) => self
+                .file
+                .write_all(&XZ_HEADER_MAGIC[..pos])
+                .await
+                .map_err(FileDownloadError::WriteError),
+            AutoXzState::PlainText => Ok(()),
+            AutoXzState::Compressed(ref mut stream, ref mut buffer) => {
+                Self::decompress(&mut self.file, stream, buffer, b"").await
+            }
+        }
+    }
+
+    /// If object is still in `Probing` state, read more data from the input to
+    /// determine whether it’s XZ stream or not.  Updates `state` accordingly.
+    /// If probing succeeded, returns number of bites from XZ header magic that
+    /// need to be processed before `chunk` is processed.  If the entire data
+    /// from `chunk` has been processed and it should be discarded by the
+    /// caller, returns `None`.
+    fn probe(&mut self, chunk: &[u8]) -> Option<usize> {
+        if chunk.is_empty() {
+            None
+        } else if let AutoXzState::Probing(pos) = self.state {
+            let len = std::cmp::min(XZ_HEADER_MAGIC.len() - pos, chunk.len());
+            if XZ_HEADER_MAGIC[pos..(pos + len)] != chunk[..len] {
+                self.state = AutoXzState::PlainText;
+                Some(pos)
+            } else if pos + len == XZ_HEADER_MAGIC.len() {
+                let stream = xz2::stream::Stream::new_stream_decoder(u64::max_value(), 0).unwrap();
+                // TODO(mina86): Once ‘new_uninit’ feature gets stabilised
+                // replaced buffer initialisation by:
+                //     let buffer = Box::new_uninit_slice(64 << 10);
+                //     let buffer = unsafe { buffer.assume_init() };
+                let buffer = vec![0u8; 64 << 10].into_boxed_slice();
+                self.state = AutoXzState::Compressed(stream, buffer);
+                Some(pos)
+            } else {
+                self.state = AutoXzState::Probing(pos + len);
+                None
+            }
+        } else {
+            Some(0)
+        }
+    }
+
+    /// Writes data to the output file.  Panics if the object is still in
+    /// probing stage.
+    async fn write_all_impl(&mut self, chunk: &[u8]) -> Result<(), FileDownloadError> {
+        match self.state {
+            AutoXzState::Probing(_) => unreachable!(),
```

Comment:
can you test 'empty' file and very short file that contains prefix of XZ_HEADER_MAGIC?
I think in such cases, when we call 'finish' - this condition will trigger.

Reply:
`write_all_impl` is not called from `finish` so that’s not the case. Added another tests case nonetheless.